lib/membrane/rtp/demuxer/jitter_buffer.ex

```diff
@@ -0,0 +1,233 @@
+            ssrc: RTP.ssrc(),
+            payload_type: RTP.payload_type(),
+            pad: Membrane.Pad.ref() | nil,
+            clock_rate: non_neg_integer() | nil,
+            latency: Membrane.Time.t() | nil,
+            initial_latency_waiting: boolean(),
+            initialization_time: Membrane.Time.t(),
+            max_latency_timer: reference() | nil,
+            timestamp_base: Membrane.Time.t(),
+            previous_timestamp: Membrane.Time.t()
+          }
+
+    @enforce_keys [
+      :ssrc,
+      :payload_type,
+      :initial_latency_waiting,
+      :initialization_time,
+      :timestamp_base,
+      :previous_timestamp
+    ]
+
+    defstruct @enforce_keys ++
+                [
+                  buffer_store: %RTP.JitterBuffer.BufferStore{},
+                  pad: nil,
+                  clock_rate: nil,
+                  latency: nil,
+                  max_latency_timer: nil
+                ]
+  end
+
+  @spec new(ExRTP.Packet.t()) :: State.t()
+  def new(packet) do
+    %State{
+      ssrc: packet.ssrc,
+      payload_type: packet.payload_type,
+      initial_latency_waiting: true,
+      initialization_time: Membrane.Time.monotonic_time(),
+      timestamp_base: packet.timestamp,
+      previous_timestamp: packet.timestamp
+    }
+  end
+
+  @spec initialize(
+          State.t(),
+          Membrane.Pad.ref(),
+          pad_options(),
+          RTP.PayloadFormat.payload_type_mapping()
+        ) :: State.t()
+  def initialize(jitter_buffer_state, pad, pad_options, payload_type_mapping) do
+    %{clock_rate: clock_rate} =
+      RTP.PayloadFormat.resolve(
+        payload_type: jitter_buffer_state.payload_type,
+        clock_rate: pad_options.clock_rate,
+        payload_type_mapping: payload_type_mapping
+      )
+
+    time_since_initialization =
+      Time.monotonic_time() - jitter_buffer_state.initialization_time
+
+    initial_latency_left = pad_options.jitter_buffer_latency - time_since_initialization
+
+    if initial_latency_left > 0 do
+      Process.send_after(
+        self(),
+        {:initial_latency_passed, jitter_buffer_state.ssrc},
+        Membrane.Time.as_milliseconds(initial_latency_left, :round)
+      )
+    end
+
+    %State{
+      jitter_buffer_state
+      | pad: pad,
+        clock_rate: clock_rate,
+        latency: pad_options.jitter_buffer_latency,
+        initial_latency_waiting: initial_latency_left > 0
+    }
+  end
+
+  @spec latency_timer_expired(State.t()) :: State.t()
+  def latency_timer_expired(jitter_buffer_state) do
+    %State{jitter_buffer_state | max_latency_timer: nil}
+  end
+
+  @spec initial_latency_passed(State.t()) :: State.t()
+  def initial_latency_passed(jitter_buffer_state) do
+    %State{jitter_buffer_state | initial_latency_waiting: false}
+  end
+
+  @spec insert_buffer(State.t(), Membrane.Buffer.t(), Demuxer.stream_phase()) :: State.t()
+  def insert_buffer(jitter_buffer_state, buffer, stream_phase) do
```

Comment:
I am not convinced we should pass `stream_phase` here - it introduces some "dependency loop" (`Demuxer` refers to `JitterBuffer` and the `JitterBuffer` refers to `Demuxer`. Perhaps that `case` clause could be part of the Demuxer's code?